perfkitbenchmarker/analyze.py

```diff
@@ -0,0 +1,202 @@
+    durations: a pd.Series of durations, as floats measured in seconds.
+    interval_start: float. The POSIX timestamp when the last stream started.
+    interval_duration: float. The time in seconds that all streams were active.
+
+  Returns: a tuple of
+    - The time between when the first and last streams started.
+    - The time between when the first and last streams ended.
+  """
+
+  interval_end = interval_start + interval_duration
+  first_start = start_times.min()
+  last_end = (start_times + durations).max()
+
+  return interval_start - first_start, last_end - interval_end
+
+
+def FullyInInterval(start_times, durations, interval_start, interval_duration):
+  """Compute which records are completely inside an interval.
+
+  Args:
+    start_times: a pd.Series of POSIX timestamps, as floats
+    durations: a pd.Series of durations in seconds, as floats
+    interval_start: the POSIX timestamp of the interval start, as a float
+    interval_duration: the duration of the interval in seconds, as a float
+
+  start_times and durations must have matching indices. Each
+  (start_time, duration) pair is considered a record of an operation
+  that began at the given start time and continued for the given
+  duration.
+
+  Returns: a pd.Series of booleans. An element is True if the
+    corresponding record lies completely within the given interval,
+    and false otherwise.
+  """
+
+  interval_end = interval_start + interval_duration
+  record_ends = start_times + durations
+
+  return (start_times >= interval_start) & (record_ends <= interval_end)
+
+
+def AllStreamsThroughputStats(durations, sizes, stream_ids,
+                              num_streams, interval_duration):
+  """Compute the net throughput of multiple streams doing operations.
+
+  Args:
+    durations: a pd.Series of durations in seconds, as floats.
+    sizes: a pd.Series of bytes.
+    stream_ids: a pd.Series of any type.
+    num_streams: int. The number of streams.
+    interval_duration: a float. The time all streams were active, in seconds.
+
+  durations, sizes, and stream_ids must have matching indices. This
+  function computes the per-stream net throughput (sum of bytes
+  transferred / total transfer time) and then adds the results to find
+  the overall net throughput. Operations from the same stream cannot
+  overlap, but operations from different streams may overlap.
+
+  Returns: a tuple of
+    - The net throughput of all streams, excluding times they were inactive.
+    - The throughput of all streams, including times they were inactive.
+    - The total time that all streams were inactive.
+    - The total time all streams were inactive, as a proportion of the
+      total benchmark time.
+  """
+
+  total_bytes_by_stream = sizes.groupby(stream_ids).sum()
+  active_time_by_stream = durations.groupby(stream_ids).sum()
+  total_overhead = interval_duration * num_streams - active_time_by_stream.sum()
```

Comment:
Is there a requirement from the caller that the "durations" array must be those durations that are covered by "interval_duration"? (i.e. when all streams are actively doing work)

If such requirement is true, then it make sense here and total_overhead must be >= 0. Please add an assertion here. (If total_overhead is negative, it doesn't make sense and something is wrong with our calculation)

Please also consider renaming total_overhead. Visually, I consider these are "gaps" in between consecutive object uploads or downloads. It's a benchmarking overhead. but in terms of analysis, these are just "gaps" in between two durations? How do you think?